### PR TITLE
Adds Reagent Conflicting

### PR DESCRIPTION
**Pull request by Iamgoofball**
:cl: Iamgoofball
experimental: To cut back on costs, Nanotrasen is removing safeguards on chemicals. Please read the labels on medication before taking it.
/:cl:

People said Chemistry has no balance besides mix time. I don't agree, but if the players demand a harder chemistry experience, who am I to deny it?

Seek alternative medication to handle dealing with bad medicine combos. Read the labels and do your homework.

Why yes, this is excessively cruel. But you know what? This is a dangerous station with dangerous experimental medication. There's a way to deal with every issue here, just open your mind.

### CODE REVIEW

**Comment by Iamgoofball** [comment]
this was done on a lunch break, will fix up any accidental compile errors when I get off work

**Comment by praisenarsie** [comment]
now remove trekchems

**Comment by Iamgoofball** [comment]
tl;dr: mixing meds with certain meds will nullify their effect and cause serious side effects

**Comment by ShizCalev** [comment]
Drug interactions, love it.

**Comment by epochayur** [comment]
if it makes medbay more complex, I approve

**Comment by ma44** [comment]
>Mix some silver sulf, styptic powder, sal acid, bicaridine and tricord
>Acquire a mix that does a total 16 damage per life cycle that can't be countered by charcoal or meds

Could maybe only have one damaging effect going on and nullify the rest, otherwise you could make some meme death chem mixes (as if there wasn't already a thing)

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 1]
```diff
@@ -891,6 +963,14 @@
 	reagent_state = LIQUID
 	color = "#C8A5DC"
 	overdose_threshold = 30
+	conflicting_reagents = list("perfluorodecalin", "salbutamol")
+
+/datum/reagent/medicine/dexalin/on_reagent_conflict(mob/living/M, datum/reagent/CR)
+	if(!..())
+		return FALSE
+	switch(CR.id)
+		if("dexalin", "perfluorodecalin")
```
Comment (ExcessiveUseOfCobblestone) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: I believe you mean salbu.

**Comment by Cyberboss** [comment]
Not sure how I feel about nullifying the original effect. I'd prefer a conflict to have some opposite, harder orthagonal effect in addition to the intended effect

**Comment by ExcessiveUseOfCobblestone** [comment]
That would make sense, yes.

**Comment by epochayur** [comment]
do brute packs/ointment actually apply any chems or is it straight up damage removal? I can see this being a real bitch if you get somebody in with multiple damage types and having to resort to either using the sleeper to inject a chem then relying on cryo to heal the rest, or smacking them with a bunch of synthiflesh

**Comment by Cyberboss** [comment]
packs/ointments just remove damage iirc

**Comment by Xhuis** [comment]
It's probably a good idea to add seeing reagents that cause issues with the ChemMaster like it displays addiction and OD info, or players have no realistic way of knowing what chemicals conflict with others

**Comment by epochayur** [comment]
Kind of going along with what xhuis said, I can see a lot of new people, for example, trying to help someone in a sleeper then accidentally dumping a bunch of conflicting chems into them, unintentionally damaging/killing the shit out of them. There needs to be a more obvious way to see what conflicts with what other than going onto git and looking at the medicine files themselves. Maybe if a sleeper detects chems in the subject that may cause a conflict when mixed with other available chems, highlight the button red but still make it click-able as a warning, just spitballing

**Comment by praisenarsie** [comment]
local man goes insane over making a billion kelotane pills

**Comment by davethwave** [comment]
Even if we don't do anything it will probably eventually become common sense for doctors and the like to figure out that so and so chemicals in someone doesn't heal them or that it may damage them. The problem is the pills that everyone takes for "safe measure" and accidentally kill themselves using the wrong medicine. Though this could actually be a good thing in my opinion as it would encourage people to head to medbay when they are injured rather than hoarding medicine. This would also be a good nerf for chems with the new chem buff that happened a few days ago here https://github.com/tgstation/tgstation/pull/34149

**Comment by 81Denton** [comment]
Can't wait for 500 accidental medbay deaths because nobody is gonna bother documenting this shit

**Comment by Iatots** [comment]
Will people still have adverse reaction when combining skin and oral applications? Example: kelotane pill + silver sulfadiazine patch

**Comment by epochayur** [comment]
^

**Comment by arsserpentarium** [comment]
Please, add non-purging toxin cures.And non healing purge medecines.
Also I'd be grateful if you'll change work of strange reagent a little. To make it revive corpse if it is in reagents, not only on applying.

**Comment by 81Denton** [comment]
Pentetic Acid and Calomel purge chems without healing

**Comment by epochayur** [comment]
>Even if we don't do anything it will probably eventually become common sense for doctors and the like to figure out that so and so chemicals in someone doesn't heal them or that it may damage them. 

a "learning curve" (i.e. frustrating trial and error for two weeks because of a decision coderbus made) is not an excuse for non-existent documentation, a player should ***never***  to have to go all the way to git to figure out what will and will not kill them when mixed together

**Comment by Iatots** [comment]
Pentetic reduces tox and rad. And calomel heals tox at <20 health.

**Comment by arsserpentarium** [comment]
Also it would be awesome if there will be shitloads of different drugs.All with side effects.But some combinations nullify negative side effects and some is worse than some poisons.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 1 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: whoops

**Comment by Iamgoofball** [comment]
I intend to make the chem-master warn about this stuff, yes.

>Not sure how I feel about nullifying the original effect. I'd prefer a conflict to have some opposite, harder, orthagonal effect in addition to the intended effect

It does make a harder effect on top of nullifying the existing healing.

**Comment by ExcessiveUseOfCobblestone** [comment]
> also because medbots are now able to murder you by accident

@Iamgoofball ensure medbots do not do dis.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 2]
```diff
@@ -379,6 +409,15 @@
 			M.reagents.remove_reagent(R.id,1)
 	..()
 
+/datum/reagent/medicine/charcoal/on_reagent_conflict(mob/living/M, datum/reagent/CR)
+	if(!..())
+		return FALSE
+	switch(CR.id)
+		if("styptic_powder", "silver_sulfadiazine", "synthflesh", "bicaridine", "kelotane", "tricordrazine")
```
Comment (cyclowns) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: couldn't it just check if its in the conflicting reagents? you already made a list for this.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 3]
```diff
@@ -194,6 +194,7 @@
 	description = "If used in touch-based applications, immediately restores burn wounds as well as restoring more over time. If ingested through other means, deals minor toxin damage."
```
Comment (kevinz000) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: All descriptions need updating unless you mean for this to be only detectable via chemmaster and codereading which is dumb.

**Comment by kevinz000** [comment]
good idea in theory but chemistry macros are still dumb and no documentation is dumb.

**Comment by Robustin** [comment]
ITT nobody reads past the title and once this gets merged we get a cacophony of "WHY DO WE KEEP LETTING GOOF GET AWAY WITH THIS?" feedback. 

We don't need months of medbay killing people just so we can say HAH THATS WHAT YOU GET FOR NOT READING CODE JACKASS THIS STATION IS DANGEROUS AND SHIT.

**Comment by Armhulen** [comment]
can we have mutation toxins conflict and turn you into a slushy mess of ??

**Comment by Iamgoofball** [comment]
yes

**Comment by Iamgoofball** [comment]
@ExcessiveUseOfCobblestone read the page, nanotrasen disabled safeguards

medibots are a crutch tool anyways lmao

**Comment by epochayur** [comment]
never change, goof

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 4]
```diff
@@ -460,7 +499,16 @@
 	color = "#D2D2D2"
 	metabolization_rate = 0.5 * REAGENTS_METABOLISM
 	overdose_threshold = 25
+	conflicting_reagents = list("synthflesh", "styptic_powder", "bicaridine")
 
+/datum/reagent/medicine/sal_acid/on_reagent_conflict(mob/living/M, datum/reagent/CR)
```
Comment (ShizCalev) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: think you meant this for omnizine

**Review by ShizCalev — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 5]
```diff
@@ -498,6 +554,14 @@
 	reagent_state = LIQUID
 	color = "#FF6464"
 	metabolization_rate = 0.25 * REAGENTS_METABOLISM
+	conflicting_reagents = list("dexalin", "salbutamol")
+
+/datum/reagent/medicine/perfluorodecalin/on_reagent_conflict(mob/living/M, datum/reagent/CR)
+	if(!..())
+		return FALSE
+	switch(CR.id)
+		if("dexalin", "perfluorodecalin")
```
Comment (ShizCalev) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: salbutamol*

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by ShizCalev** [comment]
Documentation finished and ready for launch
https://tgstation13.org/wiki/User_talk:ShizCalev

**Comment by Armhulen** [comment]
this is pretty awesome by the way amazing work @Iamgoofball

**Comment by DaedalusGame** [comment]
Why would you not use reactions for this purpose and just make the `required_container = /mob/living/carbon` and also these need to be less broad and have more unique interactions like raising or lowering bodytemp and causing hallucinations or stunning. 4/10 low-effort for a goofchem pr

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 6]
```diff
@@ -961,6 +1058,17 @@
 	color = "#C8A5DC"
 	overdose_threshold = 30
 	taste_description = "grossness"
+	conflicting_reagents = list("styptic_powder", "synthflesh", "silver_sulfadiazine", "kelotane", "bicaridine", "omnizine")
```
Comment (DaedalusGame) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: miner sticks contain omnizine and tricordrazine

**Comment by Marflowerose** [comment]
I dunno who asked for more difficult chemistry but when I do chemist, I find it enjoyable to make nice little combo pills that heal lots of stuff together. What this seems to do is just limit the production of medicine to tricord. People are still constantly ODing on shit and generally not aware that maybe that smart medicine storage has medicine, let alone the fact that eating 2 medicines for your 2 different ailments will kill you

**Comment by Magikarpet** [comment]
>inject someone with a mining survival pen
>they die

seriously though this is probably a really bad idea because a lot of things ingame stack chems intentionally and making it so that medicines are a lethal deathmix encourages people to either climb up chem's ass for mixing pills together or makes the most deadly deathmix become a mixed batch of every healing chem

**Comment by ExcessiveUseOfCobblestone** [comment]
As long as goof covers his bases and doesn't handwave every issue that arrives from his change as "well I put that they remove safeguards so #notmyproblem" then the feature will be alright I think.

**Comment by WJohn** [comment]
honestly with a system like this, you could probably remove a whole fuckload of chems in favor of chem interactions being how you perform most of the healing/damaging.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 6 above)
Comment (DaedalusGame) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Ok I admit that i was partially wrong about this because omnizine decays at one quarter the rate, 5u tricord will only neutralize 1.25u of omnizine but you- wait a second that means 5u of omnizine is enough to stay for the entire 15u of tricord so actually you take 150 damage minus what salbutamol and mining nanites heal with 1.25u of omnizine leftover

**Comment by WJohn** [comment]
actually, now that i think about this is literally kind of a pointless system to add.

you might as well have just added a bunch of new poisons (with new names) that are produced from mixing various existing healing chems. a whole "conflict" system is kind of unecessary.

**Comment by ExcessiveUseOfCobblestone** [comment]
> you might as well have just added a bunch of new poisons (with new names) that are produced from mixing various existing healing chems. a whole "conflict" system is kind of unecessary.

darn you're right.

**Comment by DaedalusGame** [comment]
Also make it less blanket and interactions more interesting, you can prevent superheal mixes without making bicard and kelo into histamine

**Comment by 81Denton** [comment]
Wouldn't this also mess with cyborg gumballs and lollipops on top of medbots, sleepers and miner pens

**Comment by DaedalusGame** [comment]
if you give a guy 1 gumball and 1 lollipop they'll take 30 brute, 10 tox damage and heal 20 burn damage
also if you give people lollipops and they have saline they'll start depleting blood
ambrosia gaia is safe
ambrosia deus is safe unless you take it with vulgaris
ambrosia vulgaris will do tox + brute + burn damage

File: code/modules/reagents/chemistry/reagents.dm [hunk 7]
```diff
@@ -112,6 +113,10 @@
 		to_chat(M, "<span class='boldannounce'>You're not feeling good at all! You really need some [name].</span>")
 	return
 
+/datum/reagent/proc/on_reagent_conflict(mob/living/M, datum/reagent/CR)
+	if(!M || !CR)
```
Comment (DaedalusGame) [inline code/modules/reagents/chemistry/reagents.dm]: on_mob_life depletes the reagent but this doesn't, conflicting reagents never go away. All my math was wrong and if you inject even 0.1u of tricord and 0.1u of omni you will take 4 damage per cycle until you stop metabolizing

can somebody confirm that this is happening or am I losing my mind

merge this please

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 6 above)
Comment (Mey-Ha-Zah) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: So, Goof will literally Kill miners with this PR.

GOOF GET THIS SHIT OUTTA HERE!!

**Comment by Repukan** [comment]
I love that the new miner pens will probably be the new sleepy pens

**Comment by Mey-Ha-Zah** [comment]
Except, Miner pens will be lethal faster then sleepypens put someone to sleep

**Comment by Iamgoofball** [comment]
Git gud

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 6 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Why do miner pens stack cure alls lmao

Ill fix that

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 3 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: Chem master only

**Comment by Iamgoofball** [comment]
Bigger question is who gave miners a pen with tricord, omni, and nanites

Fuckin hugbox lmao

**Comment by Buggy123** [comment]
I have no fucking idea on how to feel about this so I just pressed all the buttons.

On one hand, this fucks up so much stuff. On the other, this means that in effect (if I'm reading this correctly), medicines become viable and _potent_ poisons, which would be pretty interesting.

But if you dare nerf mining stimpens I swear to god...

**Comment by Xhuis** [comment]
Breaking multiple things and then telling people to "git gud" instead of saying "yes, I did break these unintentionally" is very silly

**Comment by DaedalusGame** [comment]
ah yes i intended to make a mix of bicard and kelo kill people unless they flush themselves with camo

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 6 above)
Comment (DaedalusGame) [reply code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: miner pens stack cure alls because lavaland is killzone

**Comment by Iamgoofball** [comment]
dont mix them then lmao

**Comment by DaedalusGame** [comment]
ah yes, i too love poisons that never decay and prevent you from healing yourself

**Comment by Iamgoofball** [comment]
oh boy is it "I cant rub two brain cells together and tell what is/isnt an oversight and bug" time again

**Comment by DaedalusGame** [comment]
dude it never goes away
it's better than any other poison in the game

File: code/modules/reagents/chemistry/reagents.dm (hunk 7 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/reagents/chemistry/reagents.dm]: the first check is the reagent so if you deplete either of the reagents this begins returning false.

**Comment by ExcessiveUseOfCobblestone** [comment]
> dude it never goes away

I d-d-d-don't think that's correct.

**Comment by DaedalusGame** [comment]
conflicts prevent on_mob_life from being called, which is responsible for depleting chems. the new proc doesn't deplete chems, hence conflicting reagents will never go away

File: code/modules/reagents/chemistry/holder.dm [hunk 8]
```diff
@@ -268,7 +280,17 @@
...
+				if(conflict_reagents)
+					for(var/assoc_list in conflict_reagents)
+						var/datum/reagent/base_reagent = assoc_list[1]
+						var/datum/reagent/conflict_reagent = assoc_list[2]
+						if(base_reagent == R)
+							conflicting = TRUE
+							if(base_reagent && conflict_reagent)
+								need_mob_update += R.on_reagent_conflict(C, conflict_reagent)
+				if(!conflicting)
+					need_mob_update += R.on_mob_life(C)
```
Comment (Xhuis) [inline code/modules/reagents/chemistry/holder.dm]: Actually, he's correct. unless I'm wrong. `on_mob_life` is what depletes the reagent, and it's never called if there's conflicting reagents, in lieu of calling ` on_reagent_conflict` instead. So yes, the poisons last forever.

**Comment by ExcessiveUseOfCobblestone** [comment]
hahahahaha yeah just do this WJ way where mixing medicines can make a toxic substance

**Comment by DaedalusGame** [comment]
please testmerge this anyway

File: code/modules/reagents/chemistry/holder.dm (hunk 8 above)
Comment (Iamgoofball) [reply code/modules/reagents/chemistry/holder.dm]: yes, it's a bug

**Comment by Iamgoofball** [comment]
@ExcessiveUseOfCobblestone nah m8 dont feel like letting people cheese it with antitoxins

**Comment by DaedalusGame** [comment]
then make a toxic reagent that isn't a subpath of /datum/reagent/toxin

**Comment by Iamgoofball** [comment]
Fuck off. I'm not gimping a modular conflict system because you're salty over complexity

**Comment by ExcessiveUseOfCobblestone** [comment]
is that a problem with the system or the fact charcoal "does it for free"?

**Comment by Kraseo** [comment]
MODULAR 👏 SYSTEMS 👏 AREN'T 👏 ALWAYS 👏 NECESSARY 👏 OR 👏 USEFUL 👏

**Comment by CPTANT** [comment]
This is one of those ideas that sound very nice at first.

But then you realize that people will just use this to poison people.

**Comment by Iamgoofball** [comment]
@Kraseo too bad lol

**Comment by GuyonBroadway** [comment]
Now make it so mixing neurotoxin and perflurocardine nullifies the effect of both and deals oxy damage.

**Comment by Hatfish** [comment]
so botany will become even more op as soon as they get deus, gaia, and their genetics that give omnizine and earthsblood.  wonderful

**Comment by DaedalusGame** [comment]
rofl i think mannitol and omnizine don't conflict and earthsblood and omnizine don't conflict

**Comment by Iamgoofball** [comment]
Nah

**Comment by DaedalusGame** [comment]
It is I, and I will not accept feedback that requires me to put more effort into a feature than a 30 minute lunch break

**Comment by Iamgoofball** [comment]
P. Much

**Comment by Iamgoofball** [comment]
stale
